Ideal gens merge

AlexD97 opened the pull request:
Introducing Gröbner bases/standard bases and ideal generating systems, as discussed. This PR contains the necessary amount of changes and resulting streamlining, while staying as close as possible to the status quo. Generalizations and more functionality for ideal generating systems to come as next steps.

AlexD97 committed:
IdealGens

AlexD97 committed:
IdealGens

AlexD97 committed:
Merge groebner basis as IdealGens

AlexD97 committed:
Before removing

AlexD97 committed:
IdealGens show (broken)

AlexD97 committed:
Fix IdealGens as generators

AlexD97 committed:
Merge remote-tracking branch 'upstream/master' into IdealGensMerge

ederc commented on src/Rings/groebner.jl:
```diff
@@ -61,39 +60,69 @@ returned in `B.S`.
+  complete_reduction && @assert is_global(ordering)
+  if !haskey(I.gb, ordering)
+    GB = _compute_std_basis(I.gens, ordering, complete_reduction)
+    I.gb[ordering] = GB
```
```suggestion
    I.gb[ordering] = _compute_std_basis(I.gens, ordering, complete_reduction)
```

ederc commented on src/Rings/groebner.jl:
````diff
@@ -61,39 +60,69 @@ returned in `B.S`.
+ y
+```
+"""
+function std_basis(I::MPolyIdeal, ordering::MonomialOrdering, complete_reduction::Bool = false)
````
In the other Gröbner related functions we have, in general, `ordering` as a keyword argument.

ederc commented on src/Rings/groebner.jl:
```diff
@@ -379,37 +416,8 @@ function normal_form(A::Vector{T}, J::MPolyIdeal) where { T <: MPolyElem }
 function normal_form(f::MPolyElem, J::MPolyIdeal, o::MonomialOrdering)
-  groebner_assure(J, o, false, false)
+  stdJ = std_basis(J, o, false)
   stdJ = J.gb[o]
```
Why do we need this line if `stdJ` is already set?

ederc commented on src/Rings/msolve/f4.jl:
```diff
@@ -105,7 +105,7 @@ function f4(
-    I.gb[degrevlex(vars)] = BiPolyArray(basis, keep_ordering=false, isGB=true)
+    I.gb[degrevlex(vars)] = IdealGens(basis, keep_ordering=false, isGB=true)
     
     return basis
```
In order to coincide with the return value of `std_basis` and `groebner_basis` `f4` should also return the `IdealGens` structure of the GB:
```suggestion
    return I.gb[degrevlex(vars)]
```

AlexD97 committed:
some small fixes

AlexD97 committed:
Merge remote-tracking branch 'upstream/master' into IdealGensMerge

AlexD97 committed:
IdealGens in f4

ederc commented on src/Rings/groebner.jl:
````diff
@@ -110,48 +138,48 @@ julia> H = groebner_basis(I, ordering=lex(gens(R)))
 ```
 """
-function groebner_basis(I::MPolyIdeal; ordering::MonomialOrdering = default_ordering(base_ring(I)), complete_reduction::Bool=false, enforce_global_ordering::Bool = true)
-    groebner_assure(I, ordering, complete_reduction, enforce_global_ordering)
````
Why don't we need `groebner_assure` here anymore?

AlexD97 replied:
The logic of groebner_assure is now in std_basis().

ederc commented on src/Rings/mpoly.jl:
```diff
@@ -207,12 +201,96 @@ mutable struct BiPolyArray{S}
+    r = new{T}()
+    r.gens = BiPolyArray(Ox, O)
+    r.isGB = isGB
+    r.keep_ordering = keep_ordering
```
But `ord` should be also set if the input is coming from an Oscar array.

AlexD97 replied:
You mean by an additional argument?

ederc replied:
If possible exactly like the corresponding function with Singular input. That one works without an extra argument.

AlexD97 replied:
There the ordering comes from the Singular side. But here I don’t know from where to get the ordering. 
Nevertheless, in the current state the function is wrong since there shouldn’t be any case where `isGB` is true but there is no ordering.

ederc replied:
Well, then the question is: Do we allow instantiations of `IdealGens` without `ord` being set? I would assume "no" since this is one of the main features of `IdealGens`.

jankoboehm replied:
We allow the ordering to not being set, since the data structure is also for plain generating systems. If isGB is true then an ordering has to be set. If the input comes from an OSCAR array then the ordering has to be explicitely specified. There should not be any situation where we know that something is a Gröbner basis, but do not know the respective ordering.

wdecker commented:
The elements of a GB are referred to using `gens`:
   julia> gens(groebner_basis(J))
How about replacing `gens` by `elements`?

Any thoughts somebody?

simonbrandhorst commented:
> The elements of a GB are referred to using `gens`: julia> gens(groebner_basis(J)) How about replacing `gens` by `elements`?
> 
> Any thoughts somebody?

I would say that `groebner_basis(J)` should be iterable, and  `groebner_basis(J)[1]` should work and `collect(groebner_basis(J))`.

```
julia> S  = Set([1,2,3,4])
Set{Int64} with 4 elements:
  4
  2
  3
  1

julia> collect(S)
4-element Vector{Int64}:
 4
 2
 3
 1
```
We could allow both `gens(groebner_basis(J))`  and `elements(groebner_basis(J))`. Can't hurt. Maybe one user searches for `gens` and the other likes `elements` better. This way both are happy.

AlexD97 committed:
Fix some documentation

AlexD97 committed:
Merge upstream/master

AlexD97 committed:
Fix tests

AlexD97 committed:
Fix typo

ederc committed:
Some docu test fixes

ederc committed:
more docu fixes

ederc committed:
Connects IdealGens to new f4

Adjusts support of new IdealGens interface in new f4 interface due to AlgebraicSolving

ederc committed:
Fixes f4 tests

ederc reviewed (approved):
Fine by me right now. Does anyone else have other comments resp. requests or can we merge?

tthsqe12 commented:
Let me have one another look in 12 hours.

tthsqe12 commented:
I see that that the fields `isGB`, `ord` and `keepordering` in the old `BiPolyArray` are moved to the new `IdealGens`.
Any comments on the efficacy of this two-level approach?
old:
```
BiPolyArray:
  Ox
  O
  Sx
  S
  isGB
  ord
  keepordering
```
new:
```
BiPolyArray:
  Ox
  O
  Sx
  S

IdealGens:
  gens::BiPolyArray
  isGB
  ord
  keepordering
```

ederc commented:
As far as I understood there will soon be more (meta) data added to `IdealGens` than just the ordering. I also raised the point of efficiancy some months ago but maybe @jankoboehm can give us more insight on the future plans for `IdealGens`.

tthsqe12 commented:
in any case, merge is fine with me
